envpool/mujoco/dmc/manipulator.h

```diff
@@ -0,0 +1,417 @@
+                                      data_->site_xpos[id_site_2 * 3 + 0],
+                                  data_->site_xpos[id_site_1 * 3 + 1] -
+                                      data_->site_xpos[id_site_2 * 3 + 1],
+                                  data_->site_xpos[id_site_1 * 3 + 2] -
+                                      data_->site_xpos[id_site_2 * 3 + 2]};
+    return std::sqrt(diff[0] * diff[0] + diff[1] * diff[1] + diff[2] * diff[2]);
+  }
+
+  [[nodiscard]] float IsClose(double distance) const {
+    // return rewards.tolerance(distance, (0, _CLOSE), _CLOSE * 2)
+    return static_cast<float>(RewardTolerance(distance, 0, kClose, kClose * 2));
+  }
+
+  float PegReward() {
+    float grasping = (IsClose(SiteDistance("peg_grasp", "grasp")) +
+                      IsClose(SiteDistance("peg_pinch", "pinch"))) /
+                     2;
+
+    float bringing = (IsClose(SiteDistance("peg", "target_peg")) +
+                      IsClose(SiteDistance("target_peg_tip", "peg_tip"))) /
+                     2;
+    return std::max(bringing, grasping / 3);
+  }
+  float BallReward() {
+    // return self._is_close(physics.site_distance('ball', 'target_ball'))
+    return IsClose(SiteDistance("ball", "target_ball"));
+  }
+
+  bool TaskShouldTerminateEpisode() override { return false; }
+
+ private:
+  std::array<mjtNum, 16> BoundedJointPos() {
+    std::array<mjtNum, 16> bound;
+    for (int i = 0; i < k_arm_joints_.size(); i++) {
+      int id = mj_name2id(model_, mjOBJ_JOINT, k_arm_joints_[i].c_str());
+      bound[i * 2 + 0] = std::sin(data_->qpos[id]);
+      bound[i * 2 + 1] = std::cos(data_->qpos[id]);
+    }
+    return bound;
+  }
+
+  // bug :(
+  std::array<mjtNum, 8> JointVelArm() {
+    std::array<mjtNum, 8> joint;
+    for (int i = 0; i < k_arm_joints_.size(); i++) {
+      int id = mj_name2id(model_, mjOBJ_JOINT, k_arm_joints_[i].c_str());
+      joint[i] = data_->qvel[id];
+    }
+    return joint;
+  }
+
+  std::array<mjtNum, 3> JointVelObj() {
+    std::array<mjtNum, 3> joint;
+    for (int i = 0; i < object_joints_.size(); i++) {
+      int id = mj_name2id(model_, mjOBJ_JOINT, object_joints_[i].c_str());
+      joint[i] = data_->qvel[id];
+    }
+    return joint;
+  }
+
+  std::array<mjtNum, 5> Touch() {
+    std::array<mjtNum, 5> touch;
+    for (int i = 0; i < k_touch_sensors_.size(); i++) {
+      int id = GetSensorId(model_, k_touch_sensors_[i]);
+      touch[i] = std::log1p(data_->sensordata[id]);
+    }
+    return touch;
+  }
+
+  std::array<mjtNum, 4> Body2dPose(const std::string& body_names) {
+    int id = mj_name2id(model_, mjOBJ_XBODY, body_names.c_str());
+    // self.named.data.xpos[body_names, ['x', 'z']]
+    // self.named.data.xquat[body_names, ['qw', 'qy']]
+    return {data_->xpos[id * 3 + 0], data_->xpos[id * 3 + 2],
+            data_->xquat[id * 4 + 0], data_->xquat[id * 4 + 2]};
+  }
+
+  std::string MakeManipulatorModel(const Spec& spec) {
```

Comment:
you can move this part to GetManipulatorXML(spec.config["base_path"_], spec.config["task_name"_]);